### PR TITLE
Modernize some AI loops

### PR DESCRIPTION
**Pull request by gunchleoc**
Re. #1560

### CODE REVIEW

File: src/ai/defaultai.cc [hunk 1]
```diff
@@ -318,9 +317,9 @@ void DefaultAI::think() {
 	std::sort(current_task_queue.begin(), current_task_queue.end());
 
 	// Performing tasks from temporary queue one by one
-	for (uint8_t i = 0; i < current_task_queue.size(); ++i) {
+	for (const auto& task : current_task_queue) {
```
Comment (matthiakl) [inline src/ai/defaultai.cc]:
```suggestion
	for (const SchedulerTask& task : current_task_queue) {
```

File: src/ai/defaultai.cc [hunk 2]
```diff
@@ -1693,8 +1692,8 @@ void DefaultAI::update_buildable_field(BuildableField& field) {
 	static std::set<uint32_t> unique_serials;
 	unique_serials.clear();
 
-	for (uint32_t i = 0; i < immovables.size(); ++i) {
-		const Widelands::BaseImmovable& base_immovable = *immovables.at(i).object;
+	for (const auto& imm_found : immovables) {
```
Comment (matthiakl) [inline src/ai/defaultai.cc]:
```suggestion
	for (const Widelands::ImmovableFound& imm_found : immovables) {
```

File: src/ai/defaultai_warfare.cc [hunk 3]
```diff
@@ -898,10 +898,8 @@ bool DefaultAI::check_trainingsites(uint32_t gametime) {
 	}
 
 	ts_without_trainers_ = 0;  // zeroing
-	for (std::deque<TrainingSiteObserver>::iterator site = trainingsites.begin();
-	     site != trainingsites.end(); ++site) {
-
-		if (!site->site->can_start_working()) {
+	for (const auto& observer : trainingsites) {
```
Comment (matthiakl) [inline src/ai/defaultai_warfare.cc]:
```suggestion
	for (const TrainingSiteObserver& observer : trainingsites) {
```

File: src/ai/defaultai.cc [hunk 4]
```diff
@@ -194,10 +194,9 @@ DefaultAI::DefaultAI(Widelands::Game& ggame, Widelands::PlayerNumber const pid,
 			   break;
 
 		   case Widelands::NoteShip::Action::kWaitingForCommand:
-			   for (std::deque<ShipObserver>::iterator i = allships.begin(); i != allships.end();
-			        ++i) {
-				   if (i->ship == note.ship) {
-					   i->waiting_for_command_ = true;
+			   for (auto& observer : allships) {
```
Comment (matthiakl) [inline src/ai/defaultai.cc]: https://github.com/widelands/widelands/pull/4348#discussion_r503265977
I think you should follow the same advice here
```suggestion
			   for (ShipObserver& observer : allships) {
```

File: src/ai/defaultai.cc [hunk 5]
```diff
@@ -1732,8 +1731,8 @@ void DefaultAI::update_buildable_field(BuildableField& field) {
 	any_unconnected_imm = false;
 	unique_serials.clear();
 
-	for (uint32_t i = 0; i < immovables.size(); ++i) {
-		const Widelands::BaseImmovable& base_immovable = *immovables.at(i).object;
+	for (const auto& imm_found : immovables) {
```
Comment (matthiakl) [inline src/ai/defaultai.cc]:
```suggestion
	for (const Widelands::ImmovableFound& imm_found : immovables) {
```

File: src/ai/defaultai.cc [hunk 6]
```diff
@@ -2131,10 +2130,10 @@ void DefaultAI::update_mineable_field(MineableField& field) {
 void DefaultAI::update_productionsite_stats() {
 
 	// Reset statistics for all buildings
-	for (uint32_t i = 0; i < buildings_.size(); ++i) {
-		buildings_.at(i).current_stats = 0;
-		buildings_.at(i).unoccupied_count = 0;
-		buildings_.at(i).unconnected_count = 0;
+	for (auto& bo : buildings_) {
```
Comment (matthiakl) [inline src/ai/defaultai.cc]:
```suggestion
	for (BuildingObserver& bo : buildings_) {
```

File: src/ai/defaultai.cc [hunk 7]
```diff
@@ -2191,10 +2190,9 @@ void DefaultAI::update_productionsite_stats() {
 	}
 
 	// Scale statistics down
-	for (uint32_t i = 0; i < buildings_.size(); ++i) {
-		if ((buildings_.at(i).cnt_built - buildings_.at(i).unconnected_count) > 0) {
-			buildings_.at(i).current_stats /=
-			   (buildings_.at(i).cnt_built - buildings_.at(i).unconnected_count);
+	for (auto& bo : buildings_) {
```
Comment (matthiakl) [inline src/ai/defaultai.cc]:
```suggestion
	for (BuildingObserver& bo : buildings_) {
```

File: src/ai/defaultai.cc [hunk 8]
```diff
@@ -3189,8 +3181,8 @@ bool DefaultAI::construct_building(uint32_t gametime) {
 					// +1 if any consumers_ are nearby
 					consumers_nearby_count = 0;
 
-					for (size_t k = 0; k < bo.ware_outputs.size(); ++k) {
-						consumers_nearby_count += bf->consumers_nearby.at(bo.ware_outputs.at(k));
+					for (const auto& output : bo.ware_outputs) {
```
Comment (matthiakl) [inline src/ai/defaultai.cc]:
```suggestion
					for (const Widelands::DescriptionIndex& output : bo.ware_outputs) {
```

File: src/ai/defaultai.cc [hunk 9]
```diff
@@ -6394,11 +6382,11 @@ void DefaultAI::out_of_resources_site(const Widelands::ProductionSite& site) {
 
 	const uint32_t gametime = game().get_gametime();
 
-	// we must identify which mine matches the productionsite a note reffers to
-	for (std::deque<ProductionSiteObserver>::iterator i = mines_.begin(); i != mines_.end(); ++i) {
-		if (i->site == &site) {
-			if (i->no_resources_since > gametime) {
-				i->no_resources_since = gametime;
+	// we must identify which mine matches the productionsite a note refers to
+	for (auto& mine : mines_) {
```
Comment (matthiakl) [inline src/ai/defaultai.cc]:
```suggestion
	for (ProductionSiteObserver& mine : mines_) {
```

File: src/ai/defaultai_seafaring.cc [hunk 10]
```diff
@@ -89,11 +89,11 @@ uint8_t DefaultAI::spot_scoring(Widelands::Coords candidate_spot) {
 	int32_t const tree_attr = Widelands::MapObjectDescr::get_attribute_id("tree");
 	uint16_t trees = 0;
 
-	for (uint32_t j = 0; j < immovables.size(); ++j) {
-		if (immovables.at(j).object->has_attribute(rocks_attr)) {
+	for (const auto& imm_found : immovables) {
```
Comment (matthiakl) [inline src/ai/defaultai_seafaring.cc]:
```suggestion
	for (const Widelands::ImmovableFound& imm_found : immovables) {
```

File: src/ai/defaultai_seafaring.cc [hunk 11]
```diff
@@ -177,9 +177,8 @@ bool DefaultAI::marine_main_decisions(const uint32_t gametime) {
 	}
 
 	// and now over ships
-	for (std::deque<ShipObserver>::iterator sp_iter = allships.begin(); sp_iter != allships.end();
-	     ++sp_iter) {
-		if (sp_iter->ship->state_is_expedition()) {
+	for (const auto& observer : allships) {
```
Comment (matthiakl) [inline src/ai/defaultai_seafaring.cc]:
```suggestion
	for (const ShipObserver& observer : allships) {
```

File: src/ai/defaultai_seafaring.cc [hunk 12]
```diff
@@ -344,23 +343,21 @@ bool DefaultAI::check_ships(uint32_t const gametime) {
 	while (!marine_task_queue.empty()) {
 		if (marine_task_queue.back() == kStopShipyard) {
 			// iterate over all production sites searching for shipyard
-			for (std::deque<ProductionSiteObserver>::iterator site = productionsites.begin();
-			     site != productionsites.end(); ++site) {
-				if (site->bo->is(BuildingAttribute::kShipyard)) {
-					if (!site->site->is_stopped()) {
-						game().send_player_start_stop_building(*site->site);
+			for (const auto& observer : productionsites) {
```
Comment (matthiakl) [inline src/ai/defaultai_seafaring.cc]:
```suggestion
			for (const ProductionSiteObserver& observer : productionsites) {
```

File: src/ai/defaultai_seafaring.cc [hunk 13]
```diff
@@ -344,23 +343,21 @@ bool DefaultAI::check_ships(uint32_t const gametime) {
...
 		}
 
 		if (marine_task_queue.back() == kReprioritize) {
-			for (std::deque<ProductionSiteObserver>::iterator site = productionsites.begin();
-			     site != productionsites.end(); ++site) {
-				if (site->bo->is(BuildingAttribute::kShipyard)) {
-					for (uint32_t k = 0; k < site->bo->inputs.size(); ++k) {
-						game().send_player_set_ware_priority(*site->site, Widelands::wwWARE,
-						                                     site->bo->inputs.at(k),
+			for (const auto& observer : productionsites) {
```
Comment (matthiakl) [inline src/ai/defaultai_seafaring.cc]:
```suggestion
			for (const ProductionSiteObserver& observer : productionsites) {
```

File: src/ai/defaultai_warfare.cc [hunk 14]
```diff
@@ -72,8 +72,8 @@ bool DefaultAI::check_enemy_sites(uint32_t const gametime) {
 		   game(), Widelands::Area<Widelands::FCoords>(f, (vision + 3 < 13) ? 13 : vision + 3),
 		   &immovables, Widelands::FindImmovableAttackTarget());
 
-		for (uint32_t j = 0; j < immovables.size(); ++j) {
-			if (upcast(Widelands::MilitarySite const, bld, immovables.at(j).object)) {
+		for (const auto& imm_found : immovables) {
```
Comment (matthiakl) [inline src/ai/defaultai_warfare.cc]:
```suggestion
		for (const Widelands::ImmovableFound& imm_found : immovables) {
```

File: src/ai/defaultai_warfare.cc [hunk 15]
```diff
@@ -281,10 +279,10 @@ bool DefaultAI::check_enemy_sites(uint32_t const gametime) {
 				// find militarysites near our target (radius 10) to check enemies power in region
 				map.find_immovables(game(),
 				                    Widelands::Area<Widelands::FCoords>(
-				                       map.get_fcoords(Widelands::Coords::unhash(site->first)), 10),
+				                       map.get_fcoords(Widelands::Coords::unhash(observer.first)), 10),
 				                    &immovables);
-				for (uint32_t k = 0; k < immovables.size(); ++k) {
-					const Widelands::BaseImmovable& base_immovable = *immovables.at(k).object;
+				for (const auto& im_found : immovables) {
```
Comment (matthiakl) [inline src/ai/defaultai_warfare.cc]:
```suggestion
				for (const Widelands::ImmovableFound& im_found : immovables) {
```

**Comment by matthiakl** [comment]
The first review comment isn't at the top, so here is my comment again:

> https://github.com/widelands/widelands/pull/4348#discussion_r503265977
please don't use auto for simple types like Transfer*, it makes the code harder to read

I think you should follow the same advice here

**Review by Noordfrees — APPROVED** [review]

**Comment by hessenfarmer** [comment]
@Noordfrees @gunchleoc is there any reason for not mergiung this approved PR

**Comment by gunchleoc** [comment]
I wanted to merge Time/Duration first, because that branch is a lot bigger and we now have the expected merge conflict.

**Pull request merged**